File: solutions/csharp/bowling/2/Bowling.cs
using System;
using System.Collections.Generic;
using System.Linq;

// ReSharper disable once CheckNamespace
public class BowlingGame
{
	private readonly Frames _frames = new();

	public void Roll(int pins)
		=> _ = _frames.AddRoll(pins);

	public int Score()
	{
		if (!_frames.AreComplete)
		{
			throw new ArgumentException("An incomplete game cannot be scored.");
		}

		if (_frames.ArePerfect)
		{
			return 300;
		}

		int score = 0;
		for (int frameIndex = 0; frameIndex < _frames.Count; frameIndex++)
		{
			Frame currentFrame = _frames[frameIndex];

			if (currentFrame.IsStrike)
			{
				score += 10 + GetStrikeBonus(currentFrame, frameIndex);
			}
			else if (currentFrame.IsSpare)
			{
				score += 10 + GetSpareBonus(currentFrame, frameIndex);
			}
			else
			{
				score += GetNormalFrameScore(currentFrame);
			}
		}

		return score;
	}

	private static int GetNormalFrameScore(Frame currentFrame)
		=> currentFrame.RollOne + (currentFrame.RollTwo ?? 0);

	private int GetSpareBonus(Frame currentFrame, int frameIndex)
	{
		if (currentFrame.IsLastFrame)
		{
			return currentFrame.BonusRollOne ?? 0;
		}

		Frame nextFrame = _frames[frameIndex + 1];
		return nextFrame.RollOne;
	}

	private int GetStrikeBonus(Frame currentFrame, int frameIndex)
	{
		if (_frames.IsFinalFrame(frameIndex))
		{
			return GetLastFrameBonus(currentFrame);
		}

		Frame nextFrame = _frames[frameIndex + 1];
		return nextFrame.IsStrike
			? CalculateStrikeBonus(frameIndex)
			: CalculateNonStrikeBonus(nextFrame, frameIndex);
	}

	private int CalculateStrikeBonus(int frameIndex)
		=> (frameIndex + 2) < _frames.Count
			? 10 + CalculateThirdFrameBonus(_frames[frameIndex + 2])
			: 10;

	private int CalculateNonStrikeBonus(Frame nextFrame, int frameIndex)
	{
		int bonus = GetNormalFrameScore(nextFrame);
		if ((frameIndex + 2) >= _frames.Count)
		{
			return bonus;
		}

		Frame thirdFrame = _frames[frameIndex + 2];
		if (!thirdFrame.IsStrike)
		{
			bonus += thirdFrame.RollOne + (thirdFrame.RollTwo ?? 0);
		}

		return bonus;
	}

	private int GetLastFrameBonus(Frame currentFrame)
	{
		int bonus = currentFrame.BonusRollOne ?? 0;
		if (currentFrame.BonusRollTwo != null)
		{
			bonus += currentFrame.BonusRollTwo.Value;
		}

		if (!currentFrame.IsStrike
			|| (currentFrame.BonusRollOne == null)
			|| (currentFrame.BonusRollTwo != null))
		{
			return bonus;
		}

		bonus += 10;

		if (_frames.Count > 1)
		{
			bonus += _frames[^2].RollOne;
		}

		return bonus;
	}

	private static int CalculateThirdFrameBonus(Frame thirdFrame)
		=> thirdFrame.IsStrike
			? 10
			: thirdFrame.RollOne;
}

//=======================================================================

// ReSharper disable once CheckNamespace
public record Frame
{
	private const int _strike = 10;

	private readonly int? _rollOne;

	public int RollOne
	{
		get
			=> (int)_rollOne!;
		init
			=> _rollOne = ValidateRoll(value);
	}

	public int? RollTwo { get; private set; }

	public int? BonusRollOne { get; private set; }

	public int? BonusRollTwo { get; private set; }

	public bool IsLastFrame { get; init; }

	public Frame AddRoll(int pins)
	{
		if (!IsLastFrame
			|| ((_rollOne != _strike) && !IsSpare))
		{
			RollTwo = ValidateSecondRoll(pins);
		}
		else if ((BonusRollOne == null) || IsSpare)
		{
			BonusRollOne = ValidateBonusRollOne(pins);
		}
		else
		{
			BonusRollTwo = ValidateBonusRollTwo(pins);
		}

		return this;
	}

	private int? ValidateSecondRoll(int? pins)
	{
		_ = ValidateRoll(pins);
		switch (_rollOne)
		{
			case _strike:
				throw new ArgumentException("No second rolls allowed when first roll is strike.", nameof(pins));

			default:
				if ((_rollOne + pins) > _strike)
				{
					throw new ArgumentException(
						$"First and second roll cannot add up to more than {_strike} pins.",
						nameof(pins)
					);
				}

				break;
		}

		return pins;
	}

	private int? ValidateBonusRollOne(int? pins)
		=> !IsLastFrame && (_rollOne == _strike)
			? throw new InvalidOperationException("Bonus rolls only allowed when first roll of last frame is a strike.")
			: ValidateRoll(pins);

	private int? ValidateBonusRollTwo(int? pins)
	{
		_ = ValidateRoll(pins);
		return BonusRollTwo != null
			? throw new InvalidOperationException("No more bonus rolls left.")
			: BonusRollOne == null
				? throw new InvalidOperationException("First bonus roll has not been rolled yet.")
				: BonusRollOne == _strike
					? pins
					: (BonusRollOne + pins) switch
					{
						> _strike => throw new ArgumentException(
							$"First and second bonus rolls cannot add up to more than {_strike} pins.",
							nameof(pins)
						),
						_ => pins,
					};
	}

	private static int? ValidateRoll(int? pins)
	{
		_ = pins ?? throw new ArgumentNullException(nameof(pins));
		return pins switch
		{
			< 0
				=> throw new ArgumentException("It is impossible to roll less than 0.", nameof(pins)),

			> _strike
				=> throw new ArgumentException($"It is impossible to roll more than {_strike}.", nameof(pins)),
			_
				=> pins,
		};
	}

	public bool IsStrike
		=> RollOne == _strike;

	public bool IsSpare
		=> !IsStrike
		   && RollTwo.HasValue
		   && ((RollOne + RollTwo.Value) == _strike);

	public bool IsComplete
		=> IsLastFrame
			? RollOne == _strike
				? BonusRollTwo.HasValue
				: IsSpare
					? BonusRollOne.HasValue
					: RollTwo.HasValue
			: IsStrike || RollTwo.HasValue;

	public bool IsPerfect
		=> (RollOne == _strike)
		   && (BonusRollOne == _strike)
		   && (BonusRollTwo == _strike);
}

//=======================================================================

// ReSharper disable once CheckNamespace
internal class Frames
{
	private const int _finalFrame = 10;
	private readonly List<Frame> _frames = new();

	private Frame? _lastFrame
		=> _frames.Count == _finalFrame
			? _frames[_finalFrame - 1]
			: null;

	public int Count
		=> _frames.Count;

	public Frames AddRoll(int pins)
	{
		if (AreComplete)
		{
			throw new ArgumentException("A complete game cannot receive more rolls.", nameof(pins));
		}

		Frame? lastFrame = _frames.LastOrDefault();
		if (lastFrame?.IsComplete != false)
		{
			_frames.Add(new() { RollOne = pins, IsLastFrame = _frames.Count == 9 });
		}
		else
		{
			_ = lastFrame.AddRoll(pins);
		}

		return this;
	}

	internal Frame this[int frameIndex]
		=> _frames[frameIndex];

	internal bool AreComplete
		=> _lastFrame?.IsComplete == true;

	internal bool IsFinalFrame(int frameIndex)
		=> frameIndex == (Count - 1);

	internal bool ArePerfect
		=> AreComplete
		   && _frames.All(frame => frame.IsStrike)
		   && (_lastFrame?.IsPerfect == true);
}

//=======================================================================

File: solutions/scala/bowling/2/src/main/scala/Bowling.scala
import Frame.strike
import scala.collection.mutable.ListBuffer
import scala.util.{Failure, Success, Try}

class Bowling {
    private val strike: Int = Frame.strike
    private val perfectScore: Int = 300
    private val frames: Frames = Frames()
    private var invalidState: Option[String] = None

    def roll(pins: Int): Bowling =
        if invalidState.isEmpty
        then frames.addRoll(pins) match
                case Success(_) =>
                case Failure(exception) =>
                    invalidState = Some(exception.getMessage)
        this

    def score(): Either[String, Int] =
        handleInvalidState()
            .orElse(handleIncompleteGame())
            .orElse(handlePerfectGame())
            .getOrElse(computeNormalScore())

    private def handleInvalidState(): Option[Either[String, Int]] =
        invalidState.map(Left(_))

    private def handleIncompleteGame(): Option[Either[String, Int]] =
        if !frames.areComplete
        then Some(Left("An incomplete game cannot be scored."))
        else None

    private def handlePerfectGame(): Option[Either[String, Int]] =
        if frames.arePerfect
        then Some(Right(perfectScore))
        else None

    private def computeNormalScore(): Either[String, Int] =
        Right(
            frames.foldLeft(0)(
                (acc, frame) =>
                    acc + getFrameScore(frame)
            )
        )

    private def getFrameScore(frame: Frame): Int =
        if frame.isStrike
        then strike + getStrikeBonus(frame)
        else if frame.isSpare
             then strike + getSpareBonus(frame)
             else getNormalFrameScore(frame)

    private def getNormalFrameScore(frame: Frame): Int =
        frame.getRollOne + frame.getRollTwo.getOrElse(0)

    private def getSpareBonus(frame: Frame): Int =
        if frame.isFinalFrame
        then frame.getBonusRollOne.getOrElse(0)
        else frame.nextFrame match
            case Some(nextFrame) => nextFrame.getRollOne
            case None => 0

    private def getStrikeBonus(frame: Frame): Int =
        if frame.isFinalFrame
        then getLastFrameBonus(frame)
        else {
            val nextFrame = frame.nextFrame.get
            if nextFrame.isStrike
            then calculateSecondStrikeBonus(nextFrame)
            else calculateSecondNonStrikeBonus(nextFrame)
        }

    private def getLastFrameBonus(frame: Frame): Int =
        frame.getBonusRollOne.getOrElse(0)
            + frame.getBonusRollTwo.getOrElse(0)

    private def calculateSecondStrikeBonus(secondFrame: Frame): Int =
        addNextFrameBonus(secondFrame.nextFrame, strike, calculateThirdFrameBonus)

    private def calculateSecondNonStrikeBonus(secondFrame: Frame): Int =
        val bonus = getNormalFrameScore(secondFrame)
        addNextFrameBonus(secondFrame.nextFrame, bonus, getNormalFrameScore)

    private def addNextFrameBonus(
                                     nextFrame: Option[Frame],
                                     base: Int,
                                     bonusFunction: Frame => Int
                                 ): Int =
        base + nextFrame.fold(0)(bonusFunction)

    private def calculateThirdFrameBonus(frame: Frame): Int =
        if frame.isStrike
        then strike
        else frame.getRollOne
}



class Frames extends Iterable[Frame]:
    private val totalFrameCount: Int = 10
    private val finalFrameIndex: Int = totalFrameCount - 1
    private val frames: ListBuffer[Frame] = ListBuffer.empty

    private def lastFrame: Option[Frame] =
        frames.lastOption
    private def finalFrame: Option[Frame] =
        if frames.size == totalFrameCount
        then Some(frames(finalFrameIndex))
        else None

    def addRoll(pins: Int): Try[Frames] =
        validateGameNotComplete()
            .flatMap(_ => applyRoll(pins))

    private def validateGameNotComplete(): Try[Unit] =
        if areComplete
        then Failure(new IllegalArgumentException("A complete game cannot receive more rolls."))
        else Success(())

    private def applyRoll(pins: Int): Try[Frames] =
        lastFrame match
            case None =>
                startNewFrame(pins)
            case Some(frame) if frame.isComplete =>
                startNewFrame(pins)
            case Some(frame) =>
                frame.addRoll(pins).map(_ => this)

    private def startNewFrame(pins: Int): Try[Frames] =
        val newFrame =  Frame(
            rollOne = pins,
            isFinalFrame = IsPreviousFramePenultimateFrame
        )
        lastFrame.foreach(_.nextFrame = Some(newFrame))
        frames.addOne(newFrame)
        Success(this)

    private def IsPreviousFramePenultimateFrame: Boolean =
        frames.size == totalFrameCount - 1

    def areComplete: Boolean =
        frames.size == totalFrameCount
            && frames.forall(_.isComplete)

    def arePerfect:Boolean =
        areComplete
            && frames.forall(_.isStrike)
            && finalFrame.exists(_.isPerfect)

    def apply(frameIndex: Int): Frame =
        frames(frameIndex)

    override def iterator: Iterator[Frame] =
        frames.iterator

object Frame:
    val strike: Int = 10

class Frame(rollOne: Int, val isFinalFrame: Boolean):
    def getRollOne: Int = rollOne

    var nextFrame: Option[Frame] = None

    private var rollTwo: Option[Int] = None
    def getRollTwo: Option[Int] = rollTwo
    private def setRollTwo(pins: Int): Try[Frame] =
        updateField(
            validateSecondRoll(pins),
            value => rollTwo = Some(value)
        )
    private def validateSecondRoll(pins: Int): Try[Int] =
        val validatedPins = validateRoll(pins)
        validatedPins match
            case Success(p) if isStrike =>
                Failure(new IllegalArgumentException("No second rolls allowed when first roll is strike."))
            case Success(p) if rollOne + p > strike =>
                Failure(new IllegalArgumentException(s"First and second roll cannot add up to more than $strike pins."))
            case _ =>
                validatedPins

    private var bonusRollOne: Option[Int] = None
    def getBonusRollOne: Option[Int] = bonusRollOne
    private def setBonusRollOne(pins: Int): Try[Frame] =
        updateField(
            validateBonusRollOne(pins),
            value => bonusRollOne = Some(value)
        )

    private def validateBonusRollOne(pins: Int): Try[Int] =
        val validatedPins = validateRoll(pins)
        validatedPins match
            case Success(_) if !isFinalFrame =>
                Failure(new IllegalArgumentException("Bonus rolls only in final frame."))
            case Success(_) if !(isStrike || isSpare) =>
                Failure(new IllegalArgumentException("Bonus rolls only allowed after strike or spare in final frame."))
            case other =>
                other

    private var bonusRollTwo: Option[Int] = None
    def getBonusRollTwo: Option[Int] = bonusRollTwo
    private def setBonusRollTwo(pins: Int): Try[Frame] =
        updateField(
            validateBonusRollTwo(pins),
            value => bonusRollTwo = Some(value)
        )

    private def validateBonusRollTwo(pins: Int): Try[Int] =
        val validatedPins = validateRoll(pins)
        validatedPins match
            case Success(_) if bonusRollTwo.isDefined =>
                Failure(new IllegalArgumentException("No more bonus rolls left."))

            case Success(_) if bonusRollOne.isEmpty =>
                Failure(new IllegalArgumentException("First bonus roll has not been rolled yet."))

            case Success(_) if exceedsFinalFrameBonusPinLimit(pins) =>
                Failure(
                    new IllegalArgumentException(
                        s"First and second bonus rolls cannot add up to more than $strike pins."
                    )
                )

            case other =>
                other

    private def exceedsFinalFrameBonusPinLimit(pins: Int) =
        isFinalFrame
            && bonusRollOne.exists(_ != strike)
            && bonusRollOne.exists(_ + pins > strike)

    private def updateField(validation: Try[Int], update: Int => Unit): Try[Frame] =
        validation match
            case Success(value) =>
                update(value)
                Success(this)
            case Failure(exception) =>
                Failure(exception)

    private def validateRoll(pins: Int): Try[Int] =
        pins match {
            case p if p < 0 =>
                Failure(new IllegalArgumentException("It is impossible to roll less than 0."))
            case p if p > strike =>
                Failure(new IllegalArgumentException(s"It is impossible to roll more than $strike."))
            case _ =>
                Success(pins)
        }

    def getNextFrame: Option[Frame] = nextFrame
    def setNextFrame(value: Frame): Unit = nextFrame = Some(value)

    def addRoll(pins: Int): Try[Frame] =
        if isRegularRollPhase
        then setRollTwo(pins)
        else if isFirstBonusRollPhase
             then setBonusRollOne(pins)
             else setBonusRollTwo(pins)

    private def isRegularRollPhase: Boolean =
        !isFinalFrame
            || (!isStrike && !isSpare)

    private def isFirstBonusRollPhase: Boolean =
        isFinalFrame
            && bonusRollOne.isEmpty
            && (isStrike || isSpare)

    def isStrike: Boolean =
        rollOne == strike

    def isSpare: Boolean =
        !isStrike
            && rollTwo.isDefined
            && rollTwo.exists(rollOne + _ == strike)

    def isComplete: Boolean =
        if isFinalFrame
        then if isStrike
             then bonusRollTwo.isDefined
             else if isSpare
                  then bonusRollOne.isDefined
                  else rollTwo.isDefined
        else isStrike || rollTwo.isDefined

    def isPerfect: Boolean =
        isStrike
            && bonusRollOne.contains(strike)
            && bonusRollTwo.contains(strike)

